#Bus Route and Link Extraction
with help of selenium webdriver though and by method only and i also used pandas for dataframing and time lib for loading content properly

In [1]:
from selenium import webdriver
from selenium.webdriver.common.by import By
import pandas as pd
import time



def extract_route(url):
    """Extracts route information from the provided URL."""
    driver.get(url)   
    time.sleep(5)  # Wait for the page to load
    try:
        driver.execute_script("window.scrollBy(0, 1600);")  # Scroll down to load elements
        time.sleep(2)  # Allow time for elements to load after scroll
        web_container = driver.find_element(By.CLASS_NAME, 'D117_main')
        page_tabs = web_container.find_elements(By.CLASS_NAME, 'DC_117_pageTabs')
        for page in page_tabs:
            page.click()  # Click on each page tab
            time.sleep(5)  # Wait for the page content to load
            elements = web_container.find_elements(By.CLASS_NAME, 'route_details')
            for element in elements:
                try:
                    anchor = element.find_element(By.TAG_NAME, 'a')
                    route_link['routename'].append(anchor.text)  # Store route name
                    route_link['routelink'].append(anchor.get_attribute('href'))  # Store route link
                except Exception as e:
                    pass  # Handle any exceptions that occur within the inner try block
            time.sleep(1)  # Brief pause before moving to the next set of elements
    except Exception as e:
        pass  # Handle any exceptions that occur within the outer try block

# List of URLs to extract route information from
route_dict = [
    'https://www.redbus.in/online-booking/apsrtc',
    'https://www.redbus.in/online-booking/tsrtc',
    'https://www.redbus.in/online-booking/rsrtc',
    'https://www.redbus.in/online-booking/hrtc',
    'https://www.redbus.in/online-booking/uttar-pradesh-state-road-transport-corporation-upsrtc',
    'https://www.redbus.in/online-booking/pepsu',
    'https://www.redbus.in/online-booking/pepsu-punjab',
    'https://www.redbus.in/online-booking/ksrtc-kerala',
    'https://www.redbus.in/online-booking/sikkim-nationalised-transport-snt',
    'https://www.redbus.in/online-booking/meghalaya-transport-corporation-mtc'
]


# Initialize dictionaries for storing route and bus details
driver = webdriver.Chrome()

route_link = {'routename': [], 'routelink': []}

# Extract route information from each URL in the list
for url in route_dict:
    extract_route(url)

driver.quit()  # Close the browser

# Convert the route data to a DataFrame and save it as a CSV file
route_data = pd.DataFrame(route_link)
route_data.to_csv('route_data.csv', index=False)



Route Name and Route Link Store in the *route data*

In [2]:
route_data

routename  \
0                   Vijayawada to Hyderabad   
1                   Hyderabad to Vijayawada   
2                 Kakinada to Visakhapatnam   
3                 Visakhapatnam to Kakinada   
4    Chittoor (Andhra Pradesh) to Bangalore   
..                                      ...   
274                  Shillong to Hailakandi   
275             Shillong to Ramkrishnanagar   
276                  Hailakandi to Shillong   
277             Ramkrishnanagar to Shillong   
278                    Shillong to Phulbari   

                                             routelink  
0    https://www.redbus.in/bus-tickets/vijayawada-t...  
1    https://www.redbus.in/bus-tickets/hyderabad-to...  
2    https://www.redbus.in/bus-tickets/kakinada-to-...  
3    https://www.redbus.in/bus-tickets/visakhapatna...  
4    https://www.redbus.in/bus-tickets/chittoor-and...  
..                                                 ...  
274  https://www.redbus.in/bus-tickets/shillong-to-...  
275  https://www.redbus.in/bus-tickets/shillong-to-...  
276  https://www.redbus.in/bus-tickets/hailakandi-t...  
277  https://www.redbus.in/bus-tickets/ramkrishnana...  
278  https://www.redbus.in/bus-tickets/shillong-to-...  

[279 rows x 2 columns]

that 279 link passed to the bus details extraction function, and collect the bus details

Bus Detials Extraction

In [6]:

# Initialize dictionary for storing bus details
bus_dict = {
    'route_Name':[],'busname': [],'bustype': [], 'starttime': [], 'endtime': [], 'duration': [],
    'dploc': [], 'ratings': [], 'fare': [], 'seats': [] }

# Initialize the WebDriver
driver = webdriver.Chrome()

def scroll(driver):
    """Scrolls to the bottom of the page to load dynamic content."""
    last_height = driver.execute_script("return document.body.scrollHeight")
    while True:
        driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
        time.sleep(6)
        new_height = driver.execute_script("return document.body.scrollHeight")
        if new_height == last_height:
            break
        last_height = new_height

def extract_time(name,url):
    """Extracts bus timing and other details from the provided URL."""
    driver.get(url)
    time.sleep(10)  # Wait for the page to load
    try:
        buttons = driver.find_elements(By.CLASS_NAME, 'p-left-10')
    except:
        buttons = []
    if buttons and len(buttons) > 1:
        for button in reversed(buttons):
            button.click()
            time.sleep(1)  # Allow time for content to load after click
    scroll(driver)
    time.sleep(5)  # Wait for content to load after scrolling
    web_container = driver.find_elements(By.CLASS_NAME, 'bus-item-details')
    for element in web_container:
        try:
            bus_dict["route_Name"].append(name)
            print(name, end="")
        except:
            bus_dict["route_Name"].append(name)
            print(",", end="")
        try:
            bus_dict['busname'].append(element.find_element(By.CLASS_NAME, "travels").text)
            print(element.find_element(By.CLASS_NAME, "travels").text, end="")
        except:
            bus_dict['busname'].append("Not Available")
            print(",", end="")
        try:
            bus_dict['bustype'].append(element.find_element(By.CLASS_NAME, 'bus-type').text)
            print(element.find_element(By.CLASS_NAME, 'bus-type').text, end="")
        except:
            bus_dict['bustype'].append("Not Available")
            print(",", end="")
        try:
            bus_dict['starttime'].append(element.find_element(By.CLASS_NAME, 'dp-time').text)
            print(element.find_element(By.CLASS_NAME, 'dp-time').text, end="")
        except:
            bus_dict['starttime'].append("Not Available")
            print(",", end="")
        try:
            bus_dict['endtime'].append(element.find_element(By.CLASS_NAME, 'bp-time').text)
            print(element.find_element(By.CLASS_NAME, 'bp-time').text, end="")
        except:
            bus_dict['endtime'].append("Not Available")
            print(",", end="")
        try:
            bus_dict['duration'].append(element.find_element(By.CLASS_NAME, 'dur').text)
            print(element.find_element(By.CLASS_NAME, 'dur').text, end="")
        except:
            bus_dict['duration'].append("Not Available")
            print(",", end="")
        try:
            bus_dict['dploc'].append(element.find_element(By.CLASS_NAME, 'dp-loc').text)
            print(element.find_element(By.CLASS_NAME, 'dp-loc').text, end="")
        except:
            bus_dict['dploc'].append("Not Available")
            print(",", end="")
        try:
            bus_dict['ratings'].append(element.find_element(By.CLASS_NAME, 'rating').find_element(By.TAG_NAME, 'span').text)
            print(element.find_element(By.CLASS_NAME, 'rating').find_element(By.TAG_NAME, 'span').text, end="")
        except:
            bus_dict['ratings'].append("*")  # Default rating if not found
            print(",", end="")
        try:
            bus_dict['fare'].append(element.find_element(By.CLASS_NAME, 'fare').text)
            print(element.find_element(By.CLASS_NAME, 'fare').text, end="")
        except:
            bus_dict['fare'].append("Not Available")
            print(",", end="")
        try:    
            bus_dict['seats'].append(element.find_element(By.CLASS_NAME, 'seat-left').text)
            print(element.find_element(By.CLASS_NAME, 'seat-left').text)
        except:
            bus_dict['seats'].append("Not Available")
            print(",")
        time.sleep(0.4)

# List of route links (assuming route_link is already defined and populated)
for i, link in enumerate(route_link['routelink']):
    route_name = route_link['routename'][i]
    print(f"Processing route '{route_name}' ({i + 1}/{len(route_link['routelink'])}): {link} Total Extraction:{len(bus_dict["busname"])}")
    extract_time(route_name, link)


# Close the WebDriver
driver.quit()

# Convert the route data to a DataFrame and save it as a CSV file
bus_data = pd.DataFrame(bus_dict)
bus_data.to_csv('bus_data.csv', index=False)



Processing route 'Vijayawada to Hyderabad' (1/279): https://www.redbus.in/bus-tickets/vijayawada-to-hyderabad Total Extraction:0
Vijayawada to HyderabadTSRTC - 2991RAJDHANI (A.C. Semi Sleeper)11:2017:0005h 40mPNBS-VJA2.1INR 48012 Seats available
Vijayawada to HyderabadTSRTC - 1582RAJDHANI (A.C. Semi Sleeper)12:1016:4004h 30mPNBS-VJA3.1INR 48013 Seats available
Vijayawada to HyderabadTSRTC - 40162SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)12:2018:3006h 10mPNBS-VJA2.1INR 3966 Seats available
Vijayawada to HyderabadTSRTC - 1468RAJDHANI (A.C. Semi Sleeper)12:4017:1004h 30mPNBS-VJA2.7INR 48013 Seats available
Vijayawada to HyderabadTSRTC - 1547RAJDHANI (A.C. Semi Sleeper)13:0019:1006h 10mPNBS-VJA1.7INR 4806 Seats available
Vijayawada to HyderabadTSRTC - 9945SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)13:0019:1006h 10mPNBS-VJA2.7INR 3964 Seats available
Vijayawada to HyderabadTSRTC - 18762eGARUDA13:3019:5006h 20mPNBS-VJA3.5INR 5337 Seats available
Vijayawada to HyderabadTSRTC - 40331SUPER LUXURY (NON-A

In [7]:
bus_dict
bus_data = pd.DataFrame(bus_dict)

In [8]:
bus_data

route_Name  \
0                  Vijayawada to Hyderabad   
1                  Vijayawada to Hyderabad   
2                  Vijayawada to Hyderabad   
3                  Vijayawada to Hyderabad   
4                  Vijayawada to Hyderabad   
...                                    ...   
9399          Tura (Meghalaya) to Shillong   
9400                   Shillong to Silchar   
9401                 Shillong to Karimganj   
9402                 Shillong to Karimganj   
9403  Shillong to Williamnagar (Meghalaya)   

                                            busname  \
0                                      TSRTC - 2991   
1                                      TSRTC - 1582   
2                                     TSRTC - 40162   
3                                      TSRTC - 1468   
4                                      TSRTC - 1547   
...                                             ...   
9399                                NETWORK TRAVELS   
9400  Meghalaya Transport Corporation(MTC) - 163308   
9401  Meghalaya Transport Corporation(MTC) - 163360   
9402                                NETWORK TRAVELS   
9403  Meghalaya Transport Corporation(MTC) - 163306   

                                         bustype starttime endtime duration  \
0                   RAJDHANI (A.C. Semi Sleeper)     11:20   17:00  05h 40m   
1                   RAJDHANI (A.C. Semi Sleeper)     12:10   16:40  04h 30m   
2         SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)     12:20   18:30  06h 10m   
3                   RAJDHANI (A.C. Semi Sleeper)     12:40   17:10  04h 30m   
4                   RAJDHANI (A.C. Semi Sleeper)     13:00   19:10  06h 10m   
...                                          ...       ...     ...      ...   
9399                        NON A/C Seater (2+2)     19:00   08:00  13h 00m   
9400  Tata Non A/C Seater Executive Luxury (2+2)     19:15   05:00  09h 45m   
9401  Tata Non A/C Seater Executive Luxury (2+2)     19:15   04:30  09h 15m   
9402              NON A/C Seater Push Back (2+1)     21:35   06:20  08h 45m   
9403  Tata Non A/C Seater Executive Luxury (2+2)     17:00   05:00  12h 00m   

             dploc ratings     fare               seats  
0         PNBS-VJA     2.1  INR 480  12 Seats available  
1         PNBS-VJA     3.1  INR 480  13 Seats available  
2         PNBS-VJA     2.1  INR 396   6 Seats available  
3         PNBS-VJA     2.7  INR 480  13 Seats available  
4         PNBS-VJA     1.7  INR 480   6 Seats available  
...            ...     ...      ...                 ...  
9399          Tura     3.6      500   5 Seats available  
9400   Barik Point     2.9  INR 384  22 Seats available  
9401   Barik Point     3.6  INR 371  19 Seats available  
9402      Shillong     3.7  INR 750  35 Seats available  
9403  Mawiong ISBT     3.3  INR 581   8 Seats available  

[9404 rows x 10 columns]

# DATA Cleaning

In [9]:
bus_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9404 entries, 0 to 9403
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   route_Name  9404 non-null   object
 1   busname     9404 non-null   object
 2   bustype     9404 non-null   object
 3   starttime   9404 non-null   object
 4   endtime     9404 non-null   object
 5   duration    9404 non-null   object
 6   dploc       9404 non-null   object
 7   ratings     9404 non-null   object
 8   fare        9404 non-null   object
 9   seats       9404 non-null   object
dtypes: object(10)
memory usage: 734.8+ KB


# Data Preprocessing Steps

1. **Clean and Convert Data Types**
   - **Rating, Fare, and Seats**: Ensure these columns are cleaned and converted to integer and float types as appropriate.
   - **Start Time and End Time**: Convert these columns to datetime data types.

2. **Separate Bus Name and Bus ID**
   - The `bus name` and `bus id` columns are currently combined. Separate them into distinct columns.

3. **Split Route Name**
   - **Route Name**: Split this column into two separate columns: `route_from` and `route_to`.

4. **Extract and Add Seat Type and AC Type**
   - **Seat Type and AC Type**: Extract these details from the `bus type` column and create separate columns for them.

By following these preprocessing steps, you will ensure that the data is clean, well-organized, and ready for further analysis.



In [10]:
df = bus_data.copy() # the original data copy taken
df

route_Name  \
0                  Vijayawada to Hyderabad   
1                  Vijayawada to Hyderabad   
2                  Vijayawada to Hyderabad   
3                  Vijayawada to Hyderabad   
4                  Vijayawada to Hyderabad   
...                                    ...   
9399          Tura (Meghalaya) to Shillong   
9400                   Shillong to Silchar   
9401                 Shillong to Karimganj   
9402                 Shillong to Karimganj   
9403  Shillong to Williamnagar (Meghalaya)   

                                            busname  \
0                                      TSRTC - 2991   
1                                      TSRTC - 1582   
2                                     TSRTC - 40162   
3                                      TSRTC - 1468   
4                                      TSRTC - 1547   
...                                             ...   
9399                                NETWORK TRAVELS   
9400  Meghalaya Transport Corporation(MTC) - 163308   
9401  Meghalaya Transport Corporation(MTC) - 163360   
9402                                NETWORK TRAVELS   
9403  Meghalaya Transport Corporation(MTC) - 163306   

                                         bustype starttime endtime duration  \
0                   RAJDHANI (A.C. Semi Sleeper)     11:20   17:00  05h 40m   
1                   RAJDHANI (A.C. Semi Sleeper)     12:10   16:40  04h 30m   
2         SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)     12:20   18:30  06h 10m   
3                   RAJDHANI (A.C. Semi Sleeper)     12:40   17:10  04h 30m   
4                   RAJDHANI (A.C. Semi Sleeper)     13:00   19:10  06h 10m   
...                                          ...       ...     ...      ...   
9399                        NON A/C Seater (2+2)     19:00   08:00  13h 00m   
9400  Tata Non A/C Seater Executive Luxury (2+2)     19:15   05:00  09h 45m   
9401  Tata Non A/C Seater Executive Luxury (2+2)     19:15   04:30  09h 15m   
9402              NON A/C Seater Push Back (2+1)     21:35   06:20  08h 45m   
9403  Tata Non A/C Seater Executive Luxury (2+2)     17:00   05:00  12h 00m   

             dploc ratings     fare               seats  
0         PNBS-VJA     2.1  INR 480  12 Seats available  
1         PNBS-VJA     3.1  INR 480  13 Seats available  
2         PNBS-VJA     2.1  INR 396   6 Seats available  
3         PNBS-VJA     2.7  INR 480  13 Seats available  
4         PNBS-VJA     1.7  INR 480   6 Seats available  
...            ...     ...      ...                 ...  
9399          Tura     3.6      500   5 Seats available  
9400   Barik Point     2.9  INR 384  22 Seats available  
9401   Barik Point     3.6  INR 371  19 Seats available  
9402      Shillong     3.7  INR 750  35 Seats available  
9403  Mawiong ISBT     3.3  INR 581   8 Seats available  

[9404 rows x 10 columns]

# Step 1 : **Clean and Convert Data Types**
 - Rating - int type
 - fare - remove inr and float type
 - seats - remove seats avialble and convert number
 - starttime and end time - should data and time format

In [11]:
from datetime import datetime as dt

df['fare'] = df['fare'].str.replace('INR ', '')
df["fare"] = df["fare"].apply(lambda i : float(i))


df['ratings'] = df['ratings'].str.replace('*','0')
df["ratings"] = df["ratings"].apply(lambda i : float(i))

df['starttime'] = pd.to_datetime(df['starttime'], format='%H:%M')
df['endtime'] = pd.to_datetime(df['endtime'], format='%H:%M')

df['starttime'] = df['starttime'].dt.time
df['endtime'] = df['endtime'].dt.time

df.head(3)

route_Name        busname  \
0  Vijayawada to Hyderabad   TSRTC - 2991   
1  Vijayawada to Hyderabad   TSRTC - 1582   
2  Vijayawada to Hyderabad  TSRTC - 40162   

                                  bustype starttime   endtime duration  \
0            RAJDHANI (A.C. Semi Sleeper)  11:20:00  17:00:00  05h 40m   
1            RAJDHANI (A.C. Semi Sleeper)  12:10:00  16:40:00  04h 30m   
2  SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)  12:20:00  18:30:00  06h 10m   

      dploc  ratings   fare               seats  
0  PNBS-VJA      2.1  480.0  12 Seats available  
1  PNBS-VJA      3.1  480.0  13 Seats available  
2  PNBS-VJA      2.1  396.0   6 Seats available

In [12]:
df[['Bus_Name', 'Bus_Id']] = df['busname'].str.split(' - ', expand=True)
df['Bus_Id'] = df['Bus_Id'].fillna('Private Bus')
df.loc[df['Bus_Id'] != 'Private Bus', 'Bus_Id'] = 'Government Bus'

df.head(3)

route_Name        busname  \
0  Vijayawada to Hyderabad   TSRTC - 2991   
1  Vijayawada to Hyderabad   TSRTC - 1582   
2  Vijayawada to Hyderabad  TSRTC - 40162   

                                  bustype starttime   endtime duration  \
0            RAJDHANI (A.C. Semi Sleeper)  11:20:00  17:00:00  05h 40m   
1            RAJDHANI (A.C. Semi Sleeper)  12:10:00  16:40:00  04h 30m   
2  SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)  12:20:00  18:30:00  06h 10m   

      dploc  ratings   fare               seats Bus_Name          Bus_Id  
0  PNBS-VJA      2.1  480.0  12 Seats available    TSRTC  Government Bus  
1  PNBS-VJA      3.1  480.0  13 Seats available    TSRTC  Government Bus  
2  PNBS-VJA      2.1  396.0   6 Seats available    TSRTC  Government Bus

In [13]:
df['seats'] = df['seats'].str.replace('Seats available','')
df['seats'] = df['seats'].str.replace('Seat available','')
df["seats"] = df["seats"].apply(lambda i : int(i))

df.head(3)

route_Name        busname  \
0  Vijayawada to Hyderabad   TSRTC - 2991   
1  Vijayawada to Hyderabad   TSRTC - 1582   
2  Vijayawada to Hyderabad  TSRTC - 40162   

                                  bustype starttime   endtime duration  \
0            RAJDHANI (A.C. Semi Sleeper)  11:20:00  17:00:00  05h 40m   
1            RAJDHANI (A.C. Semi Sleeper)  12:10:00  16:40:00  04h 30m   
2  SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)  12:20:00  18:30:00  06h 10m   

      dploc  ratings   fare  seats Bus_Name          Bus_Id  
0  PNBS-VJA      2.1  480.0     12    TSRTC  Government Bus  
1  PNBS-VJA      3.1  480.0     13    TSRTC  Government Bus  
2  PNBS-VJA      2.1  396.0      6    TSRTC  Government Bus

In [14]:
import re

def extract_seater(bustype):
    bustype_lower = bustype.lower()  # Convert to lowercase for consistent matching

    # Determine Seater type
    if 'seater / sleeper' in bustype_lower or 'seater/sleeper' in bustype_lower or \
       'sleeper / seater' in bustype_lower or 'sleeper/seater' in bustype_lower:
        seater = 'Seater / Sleeper'
    elif 'semi sleeper / sleeper' in bustype_lower or 'semi sleeper/sleeper' in bustype_lower or \
         'sleeper / semi sleeper' in bustype_lower or 'sleeper/semi sleeper' in bustype_lower:
        seater = 'Semi Sleeper / Sleeper'
    elif 'seater / semi sleeper' in bustype_lower or 'seater/semi sleeper' in bustype_lower or \
         'semi sleeper / seater' in bustype_lower or 'semi sleeper/seater' in bustype_lower:
        seater = 'Seater / Semi Sleeper'
    elif 'push back' in bustype_lower:
        seater = 'Push Back'
    elif 'semi sleeper' in bustype_lower:
        seater = 'Semi Sleeper'
    elif 'sleeper' in bustype_lower:
        seater = 'Sleeper'
    elif 'seater' in bustype_lower:
        seater = 'Seater'
    else:
        seater = 'Unknown'
    
    return seater

def extract_ac_type(bustype):
    bustype_lower = bustype.lower()  # Convert to lowercase for consistent matching

    # Regular expressions for matching Non-AC and AC variations
    ac_patterns = [r'\b(non[-\s]?ac)\b', r'\b(non[-\s]?a/c)\b', r'\b(non[-\s]?a\.c)\b', r'\bnon[-\s]?a\.c\b']
    non_ac_patterns = [r'\b(ac)\b', r'\ba/c\b', r'\ba\.c\b']

    ac_type = 'Unknown'  # Default value if no match is found

    # Check for Non-AC types first
    for pattern in ac_patterns:
        if re.search(pattern, bustype_lower):
            ac_type = 'Non A/C'
            break

    # Check for AC types if no Non-AC match was found
    if ac_type == 'Unknown':
        for pattern in non_ac_patterns:
            if re.search(pattern, bustype_lower):
                ac_type = 'AC'
                break

    return ac_type

# Applying the functions to your DataFrame
df['Seat_Type'] = df['bustype'].apply(extract_seater)
df['Actype'] = df['bustype'].apply(extract_ac_type)

df.head(5)


route_Name        busname  \
0  Vijayawada to Hyderabad   TSRTC - 2991   
1  Vijayawada to Hyderabad   TSRTC - 1582   
2  Vijayawada to Hyderabad  TSRTC - 40162   
3  Vijayawada to Hyderabad   TSRTC - 1468   
4  Vijayawada to Hyderabad   TSRTC - 1547   

                                  bustype starttime   endtime duration  \
0            RAJDHANI (A.C. Semi Sleeper)  11:20:00  17:00:00  05h 40m   
1            RAJDHANI (A.C. Semi Sleeper)  12:10:00  16:40:00  04h 30m   
2  SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)  12:20:00  18:30:00  06h 10m   
3            RAJDHANI (A.C. Semi Sleeper)  12:40:00  17:10:00  04h 30m   
4            RAJDHANI (A.C. Semi Sleeper)  13:00:00  19:10:00  06h 10m   

      dploc  ratings   fare  seats Bus_Name          Bus_Id     Seat_Type  \
0  PNBS-VJA      2.1  480.0     12    TSRTC  Government Bus  Semi Sleeper   
1  PNBS-VJA      3.1  480.0     13    TSRTC  Government Bus  Semi Sleeper   
2  PNBS-VJA      2.1  396.0      6    TSRTC  Government Bus     Push Back   
3  PNBS-VJA      2.7  480.0     13    TSRTC  Government Bus  Semi Sleeper   
4  PNBS-VJA      1.7  480.0      6    TSRTC  Government Bus  Semi Sleeper   

    Actype  
0       AC  
1       AC  
2  Non A/C  
3       AC  
4       AC

In [15]:
df[['h', 'm']] = df['duration'].str.split(" ", expand=True)

df['h'] = df['h'].str.replace('h', '')
df['m'] = df['m'].str.replace('m', '')
df['h'] = df['h'].str.replace('-', '')
df['m'] = df['m'].str.replace('-', '')



In [16]:
# Convert to integers
df['h'] = df['h'].apply(lambda i: int(i))
df['m'] = df['m'].apply(lambda i: int(i))

df['dur_in_min'] = (df['h']*60) + df['m']
df = df.drop(columns=['h', 'm'])

In [27]:
df = df.drop(columns="h")
df = df.drop(columns="m")


In [28]:
df.head(5)

route_Name        busname  \
0  Vijayawada to Hyderabad   TSRTC - 2991   
1  Vijayawada to Hyderabad   TSRTC - 1582   
2  Vijayawada to Hyderabad  TSRTC - 40162   
3  Vijayawada to Hyderabad   TSRTC - 1468   
4  Vijayawada to Hyderabad   TSRTC - 1547   

                                  bustype starttime   endtime duration  \
0            RAJDHANI (A.C. Semi Sleeper)  11:20:00  17:00:00  05h 40m   
1            RAJDHANI (A.C. Semi Sleeper)  12:10:00  16:40:00  04h 30m   
2  SUPER LUXURY (NON-AC, 2 + 2 PUSH BACK)  12:20:00  18:30:00  06h 10m   
3            RAJDHANI (A.C. Semi Sleeper)  12:40:00  17:10:00  04h 30m   
4            RAJDHANI (A.C. Semi Sleeper)  13:00:00  19:10:00  06h 10m   

      dploc  ratings   fare  seats Bus_Name          Bus_Id     Seat_Type  \
0  PNBS-VJA      2.1  480.0     12    TSRTC  Government Bus  Semi Sleeper   
1  PNBS-VJA      3.1  480.0     13    TSRTC  Government Bus  Semi Sleeper   
2  PNBS-VJA      2.1  396.0      6    TSRTC  Government Bus     Push Back   
3  PNBS-VJA      2.7  480.0     13    TSRTC  Government Bus  Semi Sleeper   
4  PNBS-VJA      1.7  480.0      6    TSRTC  Government Bus  Semi Sleeper   

    Actype  dur_in_min       Orgin        End  
0       AC         340  Vijayawada  Hyderabad  
1       AC         270  Vijayawada  Hyderabad  
2  Non A/C         370  Vijayawada  Hyderabad  
3       AC         270  Vijayawada  Hyderabad  
4       AC         370  Vijayawada  Hyderabad

In [29]:
df[['Orgin', 'End']] = df['route_Name'].str.split(' to ', expand=True)

In [30]:
df.to_csv("output.csv", index=False)

In [31]:
import sqlalchemy
import mysql.connector

In [35]:
engine = sqlalchemy.create_engine("mysql+mysqlconnector://admin:gopi1234@gopids.c5ic8ayqm05z.ap-south-1.rds.amazonaws.com:3306/gopids")

In [38]:
df.to_sql('redbus_web',engine)

9404